doc/survey/dynamic_graph.md

```diff
@@ -0,0 +1,269 @@
+# Differentiable Programming through Dynamic Graph
+
+The implemtation of backward lies at the heart of differentiable programming. As the dynamic graph approach gains its popularity in the past a few years, this doc surveys the backward implementations from four major dynamic graph ML frameworks: Autograd, Chainer, Pytorch and Dynet.
+
+Outline
+1. The implementation overview
+1. The design choices between four frameworks
+1. What can fluid learn from them
+
+## Implementation Overview
+
+In all four frameworks, a computation graph is built at every iteration. And There are two ways to represent the graph.
+
+### 1) Tape
+Dynet uses tape (also known as Wengert list). Consider the following code snippet of an auto-encoder.
```

Comment:
It seems a misconception of "tape" here. Tape, according to [this well-known paper](http://www.bcl.hamilton.ie/~barak/papers/toplas-reverse.pdf) in the AD field, is a "log stream" generated by running a program

> Derivatives are typically computed by recording a ‘tape’ of the computation and interpreting (or run-time compiling) a transformation of the tape played back in reverse. 
> This tape is a different kind of entity than the original program. 

but not a "representation" of the program.

I did noticed that in [this well-know tutorial](https://rufflewind.com/2016-12-30/reverse-mode-automatic-differentiation), it says

> The nodes themselves are stored in a common array (Vec<Node>) that is shared by the entire expression graph, which also acts as the allocation arena. In AD literature, this shared array is often called a tape (or Wengert list). 

This saying is very misleading, as it follows:

> The tape can be thought of as a record of all the operations performed during the evaluation of the expression, which in turn contains all the information required to compute its gradient when read in reverse.)

The problem is -- the tape IS a record, but not "can be thought of".